internal/dataplane/parser/translators/httproute.go

```diff
@@ -34,100 +43,295 @@ func TranslateHTTPRoute(route *gatewayv1beta1.HTTPRoute) []*HTTPRouteTranslation
 
 // httpRouteTranslationIndex aggregates all rules routing to the same backends group.
 type httpRouteTranslationIndex struct {
-	backendsRules map[httpBackendRefsKey]*rulesEntry
+	httpRoute *gatewayv1beta1.HTTPRoute
+	rulesMeta []httpRouteRuleMeta
 }
 
-type rulesEntry struct {
-	rules       []gatewayv1beta1.HTTPRouteRule
-	ruleNumbers []int
+func (i *httpRouteTranslationIndex) setRoute(route *gatewayv1beta1.HTTPRoute) {
+	i.httpRoute = route
+	i.extractRulesMeta(route)
 }
 
-// newHTTPRouteTranslationIndex creates a new httpRouteTranslationIndex.
-func newHTTPRouteTranslationIndex() *httpRouteTranslationIndex {
-	return &httpRouteTranslationIndex{
-		backendsRules: make(map[httpBackendRefsKey]*rulesEntry),
+func (i *httpRouteTranslationIndex) extractRulesMeta(route *gatewayv1beta1.HTTPRoute) {
+	i.rulesMeta = make([]httpRouteRuleMeta, 0)
+
+	for ruleNumber, rule := range route.Spec.Rules {
+		i.rulesMeta = append(i.rulesMeta, httpRouteRuleMeta{
+			RuleNumber: ruleNumber,
+			Rule:       rule,
+		})
 	}
 }
 
-// addRoute an HTTPRoute to the index, grouping the rules by their backendRefs.
-func (i *httpRouteTranslationIndex) addRoute(route *gatewayv1beta1.HTTPRoute) {
-	for ruleNumber, rule := range route.Spec.Rules {
-		backendRefsKey := getHTTPBackendRefsKey(rule.BackendRefs...)
-		entry, ok := i.backendsRules[backendRefsKey]
-		if !ok {
-			entry = &rulesEntry{
-				rules:       []gatewayv1beta1.HTTPRouteRule{},
-				ruleNumbers: []int{},
-			}
-			i.backendsRules[backendRefsKey] = entry
+func (i *httpRouteTranslationIndex) translate() []*KongServiceTranslation {
+	translations := make([]*KongServiceTranslation, 0)
+
+	for _, rulesByBackends := range groupRulesByBackendRefs(i.rulesMeta) {
+		// each backend refs group is a Kong service
+		kongServiceTranslation := i.translateToKongService(rulesByBackends)
+		translations = append(translations, kongServiceTranslation)
```

Comment:
Since `kongServiceTranslation` is a pointer I reckon that what is done below is basically changing what has been already put into `translations`, correct?

If that's the case I'd suggest to first do all that's necessary with `kongServiceTranslation` and append it to `translations` after. Otherwise this might be misleading.